src/coord/cartesian/Grid.ts

```diff
@@ -202,10 +200,32 @@ class Grid implements CoordinateSystemMaster {
                     }
                 }
             });
-
+            //Adjust grid.width to keep xAxis labels in dom
+            const [xAxis, yAxis] = axesList[0].isHorizontal() ? axesList : axesList.slice().reverse();
```

Comment:
I think we don't need complex logic to check the first label and last label here.

In the https://github.com/apache/echarts/blob/c8a490656f29aec8564b8142adf546058d612aab/src/coord/cartesian/Grid.ts#L193 we only adjust `x`/`width`  based on the labels of yAxis. We can also check if the union rect of xAxis labels excced the canvas size. Then adjust `x`/`width` to avoid labels overflow.

Reply:
Please correct me if I misunderstand you. The function `labelUnionRect` actually returns the largest bounding box of all labels of an axis, not the union bounding box. So I am not sure how to get the union rect of xAxis with this function. Also it would be hard to get the real size of displayed bounding box of labels because echarts seems to automatically clip labels according to axis length. That's why we just need to make sure the first/last label does not exceed the dom.

Reply:
> The function labelUnionRect actually returns the largest bounding box of all labels of an axis, not the union bounding box
The method `estimateLabelUnionRect` returns the union bounding box of all labels, not the largest.

> so it would be hard to get the real size of displayed bounding box of labels because echarts seems to automatically clip labels according to axis length

I'm not sure what do you mean by clip labels. Correct me if I'm wrong. Indeed we only pick some of the labels if there are amount of theme, especially in category axis.
https://github.com/apache/echarts/blob/4a52199842039d7fa6019ee8584d1a395fbc771a/src/coord/axisHelper.ts#L324
Perhaps the last label will also be skipped. We can add some extra logic to ensure the first label and last label are also included.

Reply:
Also the current logic of estimating `interval` and get formatted labels seems has been cap

Reply:
>The method estimateLabelUnionRect returns the union bounding box of all labels, not the largest.

The logic of function [`rect.union`](https://github.com/ecomfe/zrender/blob/67c46f39c208bb86df84c0afa63399755747f4ff/src/core/BoundingRect.ts#L48-L53) is returning the union of bounding rect. 
But here in echarts, if you trace the dataflow, every `singleRect` has `{x: 0, y: 0}`. So when you use `union` on two rects at the same position, you are just returning the larger one. That's why I say it actually returns the largest bounding rect of all labels.
https://github.com/apache/echarts/blob/4a52199842039d7fa6019ee8584d1a395fbc771a/src/coord/axisHelper.ts#L327-L338
This doesn't result in errors for now because only `x/height` and `y/width` is used in older version, and they are the same for **union rect** and **largest rect**.

However, thanks to your suggestion, I double-checked [Axis.ts](https://github.com/apache/echarts/blob/master/src/coord/Axis.ts) and found the function [`getTicksCoords`](https://github.com/apache/echarts/blob/4a52199842039d7fa6019ee8584d1a395fbc771a/src/coord/Axis.ts#L169), which returns all displayed labels' (or ticks', they bind each other) coordinates. So I rewrite the changes and now I can accurately grab the first and the last label and check whether they would exceed the dom. Please review again.

P.S. Do you mind telling me how to get the dom size in Echarts? I struggled to get dom attributes but failed.

Reply:
> But here in echarts, if you trace the dataflow, every singleRect has {x: 0, y: 0}. So when you use union on two rects at the same position, you are just returning the larger one. That's why I say it actually returns the largest bounding rect of all labels.

Thanks for correct me. You are right. 

> P.S. Do you mind telling me how to get the dom size in Echarts? I struggled to get dom attributes but failed.

You can use `api.getWidth` and `api.getHeight` to get the canvas size

Reply:
But still I think get a union bounding rect and check if the rect excced the dom is a more proper way. 

We can use the first calculated result to position the labels. Then recalculate the grid size based on the first calculated union bounding rect.

Reply:
>You can use api.getWidth and api.getHeight to get the canvas size

Thanks for the tips.
>But still I think getting a union bounding rect and check if the rect excced the dom is a more proper way.

That seems to be a general solution, but echarts would adjust how many and which label to display according to dom size. Check the behaviour below, the last label is very long but echarts chooses not to display it. If we cut the grid according to bounding rect, there would appear a huge white margin to the right of the grid, and user won't ever know why.
<img width="528" alt="discussion1#16746" src="https://user-images.githubusercontent.com/14244944/163899111-ad423c4a-799d-4202-94c6-85a88d75c564.png">
So my suggestion would still be picking the right-most/left-most displayed labels and check if they exceed the dom.

Reply:
> . If we cut the grid according to bounding rect, there would appear a huge white margin to the right of the grid, and user won't ever know why.

Indeed it is an issue that's common in the layout systems. But how will "only picking right-most label" fix it? It may still happen that the right-most label is hidden after grid size changed.

Reply:
So what I want to emphasize is right-most **DISPLAYED** label. The function `getTicksCoord` would only return the labels that would be **DISPLAYED**. Those which are automatically hidden will not be included so I can just call `getTicksCoord()[0]` to get the left-most **DISPLAYED** label and the right-most alike.

Reply:
Of course, `xAxis.inverse` is also considered. For example, left-most label will be treated as the right-most if `xAxis.inverse = true`.

Reply:
> So what I want to emphasize is right-most DISPLAYED label. The function getTicksCoord would only return the labels that would be DISPLAYED. Those which are automatically hidden will not be included so I can just call getTicksCoord()[0] to get the left-most DISPLAYED label and the right-most alike.

The union bounding rect should also be from the displayed labels .

Reply:
Hmmm, do you mean use `getTicksCoord` to find displayed labels index, find their bounding box and implement union bounding rects on them? I don't see why we need to do that. Are you worrying that some too long in-middle label would exceed the dom too?

Reply:
`getViewLabels` will return all labels that will be finally displayed on the chart. See how it is used when rendering the axis.
https://github.com/apache/echarts/blob/master/src/component/axis/AxisBuilder.ts

However the `getTicksCoord` will have a different result when user-set `axisTick.interval` and `axisLabel.interval` are different. `getTicksCoord` will use the `axisTick.interval` to skip the tick lines and `getViewLabels` will use `axisLabel.interval` to skip the coord. 

<img width="731" alt="image" src="https://user-images.githubusercontent.com/841551/163910512-7b7e482c-f2e3-43ec-8f28-e5f4a09a9ab4.png">